File: LeetCode-Java/Interleaving String.java
public class Solution {
    public boolean isInterleave(String s1, String s2, String s3) {
        int m = s1.length();
        int n = s2.length();
        if(m + n != s3.length())
            return false;
        if(m == 0) return s2.compareTo(s3) == 0;
        if(n == 0) return s1.compareTo(s3) == 0;
        boolean[][] record = new boolean[m + 1][n + 1];
        for(int i = m ;i >= 0;i --)
        {
            // use flag to terminate the iteration when no possible solution exists
            boolean tempResult = false;
            for(int j = n ;j >= 0;j --)
            {
                if(i == m && j == n)
                {   
                    record[i][j] = true;
                    tempResult = true;
                    continue;
                }
                int index = i + j;
                record[i][j] = false;
                if(i < m && s3.charAt(index) == s1.charAt(i) && record[i + 1][j])
                {
                    record[i][j] = true;
                    tempResult = true;
                    continue;
                }
                if(j < n && s3.charAt(index) == s2.charAt(j) && record[i][j + 1])
                    record[i][j] = true;
            }
            if(!tempResult) return false;
        }
        return record[0][0];
    }
}

File: Leetcode-cpp/interleaving_string.cpp
class Solution {
public:
    bool isInterleave(string s1, string s2, string s3) {
        if (s3.length() != s1.length() + s2.length()) return false;
        vector<vector<bool>> record(s1.length() + 1, vector<bool>(s2.length() + 1, true));
        for (int i = 0; i < s1.length(); ++i) {
            record[i + 1][0] = (s1[i] == s3[i] && record[i][0]);
        }
        for (int i = 0; i < s2.length(); ++i) {
            record[0][i + 1] = (s2[i] == s3[i] && record[0][i]);
        }
        for (int i = 0; i < s1.length(); ++i) {
            for (int j = 0; j < s2.length(); ++j) {
                if ((s1[i] == s3[i + j + 1] && record[i][j + 1]) || 
                    (s2[j] == s3[i + j + 1] && record[i + 1][j])) {
                    record[i + 1][j + 1] = true;
                } else {
                    record[i + 1][j + 1] = false;
                }
            }
        }
        return record[s1.length()][s2.length()];
    }
};